### PR TITLE
Feat: refactor chain state

### PR DESCRIPTION
**Pull request by DimitarTAtanasov**

### CODE REVIEW

File: src/views/explorer/components/extrinsicsList/index.tsx
Comment (nikolaicholakov25) [inline src/views/explorer/components/extrinsicsList/index.tsx:1]: image
image

containers need to take the full container height

**Review by nikolaicholakov25 — CHANGES_REQUESTED** [review]

File: src/types/block.ts [hunk 1]
```diff
@@ -15,16 +15,18 @@ interface IGenericExtrinsicMethod {
 export interface IBlockExtrinsic {
   isSigned: boolean;
   method: IGenericExtrinsicMethod;
-}
-export interface IMappedBlockExtrinsic extends IBlockExtrinsic {
   signer?: {
     Id: string;
   };
+}
+export interface IMappedBlockExtrinsic {
```
Comment (pivanov) [inline src/types/block.ts:22]: new line before will be nice :)

File: src/types/chain.ts [hunk 2]
```diff
@@ -77,3 +78,27 @@ export interface IRuntime {
...
+  blockNumber: number;
+  signer: string;
+  timestamp: number;
+  isSuccess: boolean;
+  method: string;
+  section: string;
+}
+
+interface Identity {
+  name: string | undefined;
```
Comment (pivanov) [inline src/types/chain.ts:102]: hmm some times we use `?` some times we use | undefined

File: src/utils/rpc/getBlockDetails.ts [hunk 3]
```diff
@@ -178,3 +183,108 @@ export const getBlockDetailsWithPAPI = async ({
...
+      },
+    } = extrinsic;
+
+    const isTimeStampExtrinsic = method === 'set' && section === 'timestamp';
+    if (isTimeStampExtrinsic) {
+      const _args = args as { now: string };
+      // turn the time string of type "1,451,313,413,21" into a number
+      timestamp = formatPrettyNumberString(_args?.now);
+      // only the timestamp is needed so we break the loop
+      // break;
```
Comment (pivanov) [inline src/utils/rpc/getBlockDetails.ts:247]: obsolate?

**Review by pivanov — APPROVED** [review]
r+

File: src/stores/chain/index.ts [hunk 4]
```diff
@@ -281,15 +281,41 @@ const baseStore = create<StoreInterface>()(sizeMiddleware<StoreInterface>('chain
...
+                  number,
+                  hash,
+                  timestamp,
+                  eventsLength: events.length,
+                  validator: identity.address.toString(),
+                  extrinsics: blockExtrinsics,
+                  identity,
+                };
+
+                blocksData.set(blockData.value.header.number, newBlockData);
```
[the remark this answers is not in the document]
Comment (DimitarTAtanasov) [reply src/stores/chain/index.ts:314]: it was originally implemented this way, then I decided to "refactor" it  😄

File: src/stores/chain/index.ts [hunk 5]
```diff
@@ -281,15 +281,41 @@ const baseStore = create<StoreInterface>()(sizeMiddleware<StoreInterface>('chain
...
-                }
-              });
-              set({ bestBlock: bestBlock?.number, finalizedBlock: finalizedBlock?.number });
-            })
+          Promise.allSettled(promises).then((results) => {
+            results.forEach((blockData) => {
+              if (blockData.status === 'fulfilled') {
+                const blockExtrinsics = (blockData?.value.body?.extrinsics?.slice(2) ?? []).reverse().map((extrinsic) => {
+                  const { id, blockNumber, extrinsicData, timestamp, isSuccess } = extrinsic;
+                  const { method, signer } = extrinsicData;
```
[the remark this answers is not in the document]
Comment (DimitarTAtanasov) [reply src/stores/chain/index.ts]: done

File: src/stores/chain/index.ts [hunk 6]
```diff
@@ -281,15 +281,41 @@ const baseStore = create<StoreInterface>()(sizeMiddleware<StoreInterface>('chain
...
+                    isSuccess,
+                    signer: signer?.Id ?? '',
+                    method: method.method,
+                    section: method.section,
+                  };
+                },
+                );
+
+                const { value: { header: { number, hash, timestamp, identity }, body: { events } } } = blockData;
+
```
[the remark this answers is not in the document]
Comment (DimitarTAtanasov) [reply src/stores/chain/index.ts]: done

File: src/components/modals/modalJSONViewer/index.tsx [hunk 7]
```diff
@@ -7,16 +7,15 @@ import {
...
 
 import type {
+  IBlockExtrinsic,
   IMappedBlockEvent,
-  IMappedBlockExtrinsic,
 } from '@custom-types/block';
 
 interface IModalJSONViewer extends Pick<IModal, 'onClose'> {
-  jsonData: IMappedBlockExtrinsic | IMappedBlockEvent | null;
+  jsonData: IBlockExtrinsic | IMappedBlockEvent | null;
```
[the remark this answers is not in the document]
Comment (DimitarTAtanasov) [reply src/components/modals/modalJSONViewer/index.tsx]: done

File: src/stores/chain/index.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (DimitarTAtanasov) [reply src/stores/chain/index.ts]: done

File: src/types/block.ts (hunk 1 above)
Comment (DimitarTAtanasov) [reply src/types/block.ts:23]: done

File: src/types/chain.ts (hunk 2 above)
Comment (DimitarTAtanasov) [reply src/types/chain.ts]: done

File: src/utils/rpc/getBlockDetails.ts (hunk 3 above)
Comment (DimitarTAtanasov) [reply src/utils/rpc/getBlockDetails.ts]: done

**Review by nikolaicholakov25 — APPROVED** [review]

**Pull request merged**